query/src/main/java/org/infinispan/query/concurrent/InfinispanIndexingExecutorProvider.java

```diff
@@ -0,0 +1,70 @@
+package org.infinispan.query.concurrent;
+
+import static org.infinispan.factories.KnownComponentNames.BLOCKING_EXECUTOR;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.hibernate.search.backend.lucene.work.spi.LuceneWorkExecutorProvider;
+import org.hibernate.search.engine.common.execution.spi.SimpleScheduledExecutor;
+import org.infinispan.factories.AbstractComponentFactory;
+import org.infinispan.factories.AutoInstantiableFactory;
+import org.infinispan.factories.KnownComponentNames;
+import org.infinispan.factories.annotations.ComponentName;
+import org.infinispan.factories.annotations.DefaultFactoryFor;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.factories.scopes.Scope;
+import org.infinispan.factories.scopes.Scopes;
+
+@Scope(Scopes.GLOBAL)
+@DefaultFactoryFor(classes = LuceneWorkExecutorProvider.class)
+public class InfinispanIndexingExecutorProvider extends AbstractComponentFactory implements AutoInstantiableFactory, LuceneWorkExecutorProvider {
+
+   @Inject
+   @ComponentName(KnownComponentNames.TIMEOUT_SCHEDULE_EXECUTOR)
+   ScheduledExecutorService scheduledExecutorService;
+
+   @Inject
+   @ComponentName(BLOCKING_EXECUTOR)
+   ExecutorService blockingExecutor;
+
+   @Override
+   public Object construct(String name) {
+      return new InfinispanIndexingExecutorProvider();
+   }
+
+   @Override
+   public SimpleScheduledExecutor writeExecutor(Context context) {
+      return new InfinispanScheduledExecutor(scheduledExecutorService, blockingExecutor);
+   }
+
+   static final class InfinispanScheduledExecutor implements SimpleScheduledExecutor {
+
+      private final ScheduledExecutorService scheduledExecutorService;
+      private final ExecutorService blockingExecutor;
+
+      public InfinispanScheduledExecutor(ScheduledExecutorService scheduledExecutorService,
+                                         ExecutorService blockingExecutor) {
+         this.scheduledExecutorService = scheduledExecutorService;
+         this.blockingExecutor = blockingExecutor;
+      }
+
+      @Override
+      public Future<?> submit(Runnable task) {
+         return blockingExecutor.submit(task);
+      }
+
+      @Override
+      public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
+         return scheduledExecutorService.schedule(() -> blockingExecutor.submit(command), delay, unit);
```

Comment:
The more I think about it, this actually isn't sufficient. We need to create a wrapper ScheduledFuture class otherwise the Future methods won't work properly.

Reply:
Might be easier if I implement something within the BlockingManager instead so the code doesn't live here.

I am thinking `BlockingManager#scheduleBlockingRunnable(ScheduledExecutorService schedule, Runnable command, long delay, TimeUnit unit)` and then we can do the work of translating. We could also just encapsulate the scheduled executor service in the BlockingManager which then not require the argument.

Reply:
You can use the new method in https://github.com/infinispan/infinispan/pull/10610

Reply:
It isn't sufficient because the returned Future would have been completed when the task was submitted. This new way it isn't completed until the task that was submitted to the blocking thread was complete.

Reply:
Is this because we didn't use (wait for the completion of) the future returned from the `blockingExecutor#submit`?
``` java
scheduledExecutorService.schedule(() -> blockingExecutor.submit(command), delay, unit);
```

Reply:
Yes, but we can't wait for it to complete in the scheduled executor as it would tie up its thread.

Reply:
Always for curiosity couldn't we have done something like:
``` java
scheduledExecutorService.schedule(() -> blockingExecutor.submit(task).get(), delay, unit);
```

Reply:
Maybe not to correctly handle the exception case...

Reply:
No, we can't do that as it would tie up the scheduler thread that I was referencing in my previous comment. The goal is to schedule the execution on the blocking thread but not block the thread that is submitting it.

A simple rule of thumb is you should almost never be calling `get` on a Future in ISPN nowadays. If you do it _has_ to be on a blocking thread.

Reply:
Right! We should to free the scheduler immediately. Finally got it :) Thanks for the explanations